### DISCUSSION TITLE
Ctrl+C to cancel while running a long running C# method

### DISCUSSION THREAD

**Discussion by drache42** [pythonnet/pythonnet#2309] [Q&A] [answered]
Hello,
I have a question on how to cancel a long running C# method that I called from python.

Ctrl+C does not seem to get picked up either by python or by dotnet. 

I have access to the [CancellationTokenSource ](https://learn.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource?view=net-8.0) that I'm using, so if I can catch the Ctrl+C in python, then I could cancel the C#.

Thank you!

**Comment by drache42** [comment, marked as answer]
One solution I came up with was to put calling the C# method from a thread. Then I can do a busy-wait look on the thread join. Something like this

```
    thread = Thread(target = <dotnet method that is long running>)
    thread.start();

    while True:
        try:
            thread.join(1) #This will wait for 1 second which then allows the signal handler to catch the Ctrl+C
        except KeyboardInterrupt:
            print("Caught Ctrl+C, shutting down...")
            <Call the the CancellationTokenSource.Cancel() method>
```

**Comment by DanielSparta** [comment]
Correct me if I'm wrong, You created a thread at C# and you want to stop this thread running?

**Reply by drache42** [reply]
Not quite. I have a long running method in C# (not a thread).

When I call this long running C# method from python, then python does not pick up the `SIGINT` (Ctrl+C).

The fix I was able to come up with, is the code above.
I make sure the C# method is cancellable via [Cancellation Token](https://learn.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken?view=net-8.0). 
I also make sure that the python code has access to the [Cancellation Token Source](https://learn.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource?view=net-8.0) that produced the cancellation token. 

Then I can start a Python thread that executes my C# method.

The busy-wait (`while true`) loop then just waits for the thread to finish via the `.join` function. I give it a 1 second time out which will allow the signal handler to execute which in turn means the except block will execute. 
In the `except KeyboardInterrupt:` I can then call the cancellation token source `.cancel()` method. This then tells the C# method to finish what's it's doing and exit gracefully.
This in turns allows the thread to finish and then the `thread.join()` is able to gracefully return control back to the python code.

**Reply by DanielSparta** [reply]
You dont need a cancellation token to stop method run, If your inside a while(true) loop, you can make a other condition or a "global" value such as private bool run = true, and then while (run) { .... } Should also do the work.

**Reply by drache42** [reply]
If you don't use the cancellation token, then it won't stop the C# method gracefully. 
Also, if there is a global bool to determine if it runs, and it's set to false, then the thread won't actually quit. It would continue to run with killing.
Of course, you could kill the thread after run is set to false, but then again, it would not gracefully stop the C# method.

Either way, the initial issue was that the SIGINT isn't handled when running a C# method, something like the above code is a way to do it.

**Reply by DanielSparta** [reply]
C# have a Garbage collector that will close the thread for you after It's run stopped.
Please send your method code.

**Reply by drache42** [reply]
Yes, the memory will get collected. 
Gracefully means that the C# method will stop itself, to allow it to do cleanup for whatever it's doing. There are many reasons you want to be able to stop a method gracefully from running without killing it. 

Killing != stopping gracefully.

**Reply by drache42** [reply]
> You dont need a cancellation token to stop method run, If your inside a while(true) loop, you can make a other condition or a "global" value such as private bool run = true, and then while (run) { .... } Should also do the work.

> C# have a Garbage collector that will close the thread for you after It's run stopped.
Please send your method code.

Nothing in your code kills the thread, so it won't actually stop the thread (which is a python thread) unless the python application itself stops.

**Reply by DanielSparta** [reply]
When your thread code comes to end, The garbage collector will dispose it and the resources will back.
I don't know what your trying to do, but if your thread "alive" even after his code "end", and its making bugs at your code, Then probably the problem is your code, Thread that finished should not be problems

**Reply by drache42** [reply]
My original question was "How do I catch a Ctrl+C" (which is used to cancel a running application via the `SIGINT` signal). 
I have come up with a solution for this as described in my first follow up post.

In my particular example, I needed that Ctrl+C to gracefully shut down the ongoing C# method which my example code does.

If you have another way to do this, then please post your example code here so we can discuss in detail.
The details of the C# method are not relevant, only that it runs for a long time and optionally that it needs the ability to stop gracefully. If you don't need the method (and hence the python thread) to end gracefully, then killing the thread is an option.

So far this entire thread has going around in circles. I unfortunately don't have time to discuss this right now.

If you have some example python code that can achieve the objective described, then please post it and I will review it.
